### PR TITLE
[improve][admin] PIP-369 Introduce `unload` flag in `ns-isolation-policy set` call

### PR DESCRIPTION
**Pull request by grssam**
Implementation PR for the PIP 369 "Flag based selective unload on changing ns-isolation-policy" #23116 

### Modifications

Added an `unload-scope` flag in the `ns-isolation-policy set` call to control the behavior of which namespaces to unload as part of the set call in a more granular manner.

As this PR is targetted for 3.0 LTS as well, keeping the changes backward compatible. i.e. the flag `unload-scope` is optional. Default value is `all_matching` which means all namespaces matching the namespaces regex being set in the policy set call.

There will be a followup PR which will be adding the following breaking changes:

- Making the default value of the flag to be `changed` i.e. only the newly added or now-removed namespace regex would be used to unload the namespaces
- The meaning of `all_matching` will change to cover all namespaces matching both the previous and the new namespace regex values. The idea is that if immediate placement correction is needed, we should also unload the namespaces which do not belong to the isolation group anymore.

### Verifying this change

This change added tests and can be verified as follows:

  - Added and made changes to integration tests covering this feature. All three unload scopes are tried and validation is done that only the relevant namespaces are unloaded

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [x] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [x] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

PR in forked repo - https://github.com/grssam/pulsar/pull/1

### CODE REVIEW

**Comment by grssam** [comment]
Please add labels for 3.0.7 release as I would like this to be present in the LTS version.

**Review by lhotari — APPROVED** [review]
LGTM. Outstanding work @grssam! Thanks for contributing.

**Comment by grssam** [comment]
Not able to rerun the failed CI pipeline, but all tests passed in my repo - https://github.com/grssam/pulsar/pull/1 and the new tests that I added are part of "Broker Group 3" only.

**Comment by lhotari** [comment]
> Not able to rerun the failed CI pipeline, but all tests passed in my repo - https://github.com/grssam/pulsar/pull/1 and the new tests that I added are part of "Broker Group 3" only.

In Pulsar CI, you can trigger a rerun by adding a comment "/pulsarbot rerun-failure-checks". It requires that a possible previous workflow run has completed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java
Comment (grssam) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]: Good catch! I've updated it.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java
Comment (dao-jun) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java:25]: Please avoid star imports

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java
Comment (dao-jun) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java:114]: Please avoid star imports

**Review by dao-jun — APPROVED** [review]
LGTM, just a few minor comments.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java
Comment (grssam) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java:25]: Any reason why a star import is bad here? there can't be any conflict as the enum would only be usable in a place where it is expecting an enum of class `NamespaceIsolationPolicyUnloadScope`.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java
Comment (grssam) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java:114]: Any documentation around the expected order of imports? When I added my import for the added class in lexical order, the checkstyle failed with an error that my import is out of place

Comment (dao-jun) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java:114]: It's the checkstyle rule, however, the rule only applies to the source codes(non-test codes). Feel free to ignore these two comments if you are in trouble resolving them.

Comment (grssam) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java:114]: Since It is a test file, i will let this as is :)

**Comment by grssam** [comment]
PS: while cherry picking to branch-3.0 , it will contain one conflict due to the command line argument parser library change at line https://github.com/apache/pulsar/pull/23120/files#diff-63a2a931cc53a14767f5eb4a53898d7b6fb20f81a87de51cc15bbc139782bd00R77

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java
Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java:25]: @grssam The main reason is that Pulsar code base uses such a code style. You get the correct formatting in IntelliJ if you apply the Pulsar code style as instructed in https://pulsar.apache.org/contribute/setup-ide/.

Comment (grssam) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java:25]: @dao-jun Thank you for the review.. but since this is an enum import and i am using all of the enums in the tests.. I'll leave it as is.
I do understand normal class imports via star being harmful to future naming conflicts, but this one shouldn't have any even in future.

**Comment by lhotari** [comment]
> PS: while cherry picking to branch-3.0 , it will contain one conflict due to the command line argument parser library change at line https://github.com/apache/pulsar/pull/23120/files#diff-63a2a931cc53a14767f5eb4a53898d7b6fb20f81a87de51cc15bbc139782bd00R77 (and the relevant class imports as well)

@grssam  Yes, that's common that there are conflicts. The committer performing the cherry-picking and backporting will request a separate PR for a backport to branch-3.0 if there's a need. Some of this is explained in the mailing list message https://lists.apache.org/thread/ryyfv8o33yyw9hmo5gnxn71boocn3mzs .

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java:816]:
```suggestion
                List<String> oldBrokers = new ArrayList<>(oldPolicy.getPrimary());
```

It should be the primary of the old policy.

Comment (grssam) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]: Whoops. Thank you for catching this. Accepted the change.

**Review by nodece — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java
Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]: @grssam any chance to add a test to cover that part?

Comment (grssam) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]: @lhotari added. Will need approval again i guess.

**Review by lhotari — APPROVED** [review]

**Comment by dao-jun** [comment]
merging...

**Pull request merged**
